src/utils/daos.c

```diff
@@ -0,0 +1,813 @@
+	if (str != NULL && str[0] == '-') {
+		fprintf(stderr, "WARNING bytes < 0 (string %s)"
+				"converted to "DF_U64" : using 0 instead\n",
+				str, size);
+		size = 0;
+		return size;
+	}
+
+	/** no suffix used */
+	if (*end == '\0')
+		return size;
+
+	/** let's be permissive and allow MB, Mb, mb ...*/
+	if (*(end + 1) != '\0' &&
+	    ((*(end + 1) != 'b' && *(end + 1) != 'B') || (*(end + 2) != '\0')))
+		return 0;
+
+	switch (*end) {
+	case 'b':
+	case 'B':
+		break;
+	case 'k':
+	case 'K':
+		size <<= 10;
+		break;
+	case 'm':
+	case 'M':
+		size <<= 20;
+		break;
+	case 'g':
+	case 'G':
+		size <<= 30;
+		break;
+	case 't':
+	case 'T':
+		size <<= 40;
+		break;
+	case 'p':
+	case 'P':
+		size <<= 50;
+		break;
+	case 'e':
+	case 'E':
+		size <<= 60;
+		break;
+	default:
+		return 0;
+	}
+
+	return size;
+}
+
+
+static int
+epoch_range_parse(struct cmd_args_s *ap)
+{
+	int		rc;
+	long long int	parsed_begin = 0;
+	long long int	parsed_end = 0;
+
+	rc = sscanf(ap->epcrange_str, "%lld-%lld",
+			&parsed_begin, &parsed_end);
+	if ((rc != 2) || (parsed_begin < 0) || (parsed_end < 0)) {
+		D_GOTO(out_invalid_format, -1);
+	}
+
+	ap->epcrange_begin = parsed_begin;
+	ap->epcrange_end = parsed_end;
+
+	return 0;
+
+out_invalid_format:
+	fprintf(stderr, "epcrange=%s must be in A-B form\n",
+		ap->epcrange_str);
+	return -1;
+}
+
+static int
+common_op_parse_hdlr(int argc, char *argv[], struct cmd_args_s *ap)
+{
+	struct option		options[] = {
+		{"group",	required_argument,	NULL,	'G'},
+		{"pool",	required_argument,	NULL,	'p'},
+		{"svc",		required_argument,	NULL,	'm'},
+		{"cont",	required_argument,	NULL,	'c'},
+		{"attr",	required_argument,	NULL,	'a'},
+		{"value",	required_argument,	NULL,	'v'},
+		{"path",	required_argument,	NULL,	'd'},
+		{"type",	required_argument,	NULL,	't'},
+		{"oclass",	required_argument,	NULL,	'o'},
+		{"chunk_size",	required_argument,	NULL,	'z'},
+		{"snap",	required_argument,	NULL,	's'},
+		{"epc",		required_argument,	NULL,	'e'},
+		{"epcrange",	required_argument,	NULL,	'r'},
+		{NULL,		0,			NULL,	0}
+	};
+	int			rc;
+	const int		RC_PRINT_HELP = 2;
+	const int		RC_NO_HELP = -2;
+	char			*cmdname = NULL;
+
+	assert(ap != NULL);
+	ap->p_op = -1;
+	ap->c_op = -1;
+	ap->group = default_group;
+
+	if ((strcmp(argv[1], "container") == 0) ||
+	    (strcmp(argv[1], "cont") == 0)) {
+		ap->c_op = cont_op_parse(argv[2]);
+		if (ap->c_op == -1) {
+			fprintf(stderr, "invalid container command: %s\n",
+				argv[2]);
+			fflush(stderr);
+			return RC_PRINT_HELP;
+		}
+	} else if (strcmp(argv[1], "pool") == 0) {
+		ap->p_op = pool_op_parse(argv[2]);
+		if (ap->p_op == -1) {
+			fprintf(stderr, "invalid pool command: %s\n",
+				argv[2]);
+			fflush(stderr);
+			return RC_PRINT_HELP;
+		}
+	} else {
+		/* main() may catch error. Keep this code just in case. */
+		fprintf(stderr, "resource (%s): must be "
+				 "pool or container\n", argv[1]);
+		fflush(stderr);
+		return RC_PRINT_HELP;
+	}
+	cmdname = strdup(argv[2]);
```

Comment:
Probably a good idea to get all memory management stuff to be done through "gurt" utility methods. Other than malloc and free, "gurt" (cart/src/include/gurt/common.h) also provides wrappers for strdup/strndup (D_STRNDUP) and asprintf (D_ASPRINTF) and realpath (D_REALPATH).